RONDB-746: Add a thread context to ensure that each thread uses its own Ndb object

mronstro opened the pull request.

mikaelronstrom committed:
RONDB-746: Adding threadIndex parameter to function that gets and returns Ndb object

Each thread will have its own Ndb object, at the same time the object needs to be
protected such that shutdown can go reasonably fast.

mikaelronstrom committed:
RONDB-746: Added a ThreadContext to handle shutdowns better

mikaelronstrom committed:
RONDB-746: Each thread use its own Ndb object

svenssonaxel commented on storage/ndb/rest-server2/server/src/rdrs_rondb_connection_pool.cpp:
```diff
@@ -24,16 +24,48 @@
+ThreadContext::ThreadContext() {
+  m_thread_context_mutex = NdbMutex_Create();
+  require(m_thread_context_mutex);
+  m_is_shutdown = false;
```
These three assignments could be done before the body e.g. `() : m_is_shutdown(false) ... {`

mronstro replied:
Yep, but in my view this variant is easier to read, the assignment before the function simply introduces syntactical sugar that in my opinion has no real purpose, I am sure someone disagrees :)

svenssonaxel commented on storage/ndb/rest-server2/server/src/rdrs_rondb_connection_pool.cpp:
```diff
@@ -91,16 +129,61 @@ RS_Status RDRSRonDBConnectionPool::AddMetaConnections(
+    NdbMutex_Unlock(m_thread_context[threadIndex]->m_thread_context_mutex);
+    return RS_OK;
+  }
+  require(thread_context->m_is_ndb_object_in_use == false);
```
If this is correct then why not move it to the top of the function?

mronstro replied:
One could do so, yep

svenssonaxel commented:
Do I understand correctly that an NDB object represents a connection, and each NDB object would allocate one node id? If so, requiring one NDB object per thread would mean requiring one node id per thread. Is that necessary? Wouldn't it risk exhausting the node ids?

mikaelronstrom committed:
RONDB-746: First stab at adding support for multiple cluster connections for data traffic

mikaelronstrom committed:
RONDB-746: Added a bit more reporting on benchmark runs continously

mikaelronstrom committed:
RONDB-746: Benchmark fixes

mikaelronstrom committed:
RONDB-746: More checks at startup of proper config of rdrs2

Changed default of MGM server port to 1186
Allowed up to 8 cluster connections for data
Allowed 0 metadata connections in which case the first data
connection is used.
Allowed also 1 metadata connections, but not more than 1.
Allowed zero node ids or number of node ids equal to the number of
data connections and same for metadata connections.
Ensured that metadata connection isn't deleted if it is pointing to
a data connection.
Removed checks of node ids from client test programs.
More printouts when failing to start after wrong config.
Comment that Reconnect for all cluster connections not used currently.
Ensured that test programs cannot run with gRPC since it isn't supported.

mikaelronstrom committed:
RONDB-717, RONDB-718: Ping, health, stat endpoints

Ping endpoint working
Health endpoint developed
Stat endpoint developed, gathering stats TODO
